lib/IDE/SwiftSourceDocInfo.cpp

```diff
@@ -152,6 +156,14 @@ bool CursorInfoResolver::walkToStmtPre(Stmt *S) {
   // non-implicit Stmts (fix Stmts created for lazy vars).
   if (!S->isImplicit() && !rangeContainsLoc(S->getSourceRange()))
     return false;
+
+  if (isa<BraceStmt>(S)) {
+    ContextStack.push_back(ContextNode);
```

Comment:
We only capture explicit brace statement as the context node. I think the main benefit is we can start walk from the brace statement instead of the source file. However, walking AST is fairly efficient operation. I think we can afford to start walking from SourceFIle.

Reply:
OK, I will try walking from `SourceFIle`.
Can I add `SourceFile` property to `ResolvedCursorInfo` then?
Should I pass `SF` from `collectAvailableRefactorings()` to `isApplicable()`?

Reply:
That's a good idea! can we add a `SoureFile` filed to `ResolvedCursorInfo` instead of passing it? I bet other refactorings may need this information too.

Reply:
Actually mine needs it, too :) I've  changed the `isApplicable` to expect `ApplicabilityContext`, which is a struct that contains `DiagnosticEngine` reference and `SourceFile` pointer. You could check it here: https://github.com/apple/swift/pull/12281. To be honest either way works, so I'm happy to change it.
One thought, though: If we're passing `SourceFile` to the `ResolvedCursorInfo`— shouldn't we pass `DiagnosticEngine` the same way for consistency?

Reply:
> If we're passing `SourceFile` to the `ResolvedCursorInfo` — shouldn't we pass `DiagnosticEngine` the same way for consistency?

I don't think `DiagnosticEngine` should be a property of `ResolvedCursorInfo`, whereas `SourceFile` is directly related to the cursor position.

Reply:
Yeah, you're right 👍